simde/arm/neon/cmla.h

```diff
@@ -0,0 +1,371 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright:
+ *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+ */
+
+ #if !defined(SIMDE_ARM_NEON_CMLA_H)
+ #define SIMDE_ARM_NEON_CMLA_H
+
+ #include "types.h"
+
+ HEDLEY_DIAGNOSTIC_PUSH
+ SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+ SIMDE_BEGIN_DECLS_
+
+ /* * rot 0 * */
+
+ SIMDE_FUNCTION_ATTRIBUTES
+ simde_float32x2_t
+ simde_vcmla_f32(simde_float32x2_t r, simde_float32x2_t a, simde_float32x2_t b) {
+   #if defined(SIMDE_ARM_NEON_A32V8_NATIVE)
+    return vcmla_f32(r, a, b);
+   #else
+    simde_float32x2_private
+      r_ = simde_float32x2_to_private(r),
+      a_ = simde_float32x2_to_private(a),
+      b_ = simde_float32x2_to_private(b);
+
+    SIMDE_VECTORIZE
+    for(size_t i = 0 ; i < (sizeof(r_.values) / sizeof(r_.values[0])) ; i++) {
+      if(i % 2 == 0)
+        r_.values[i] += b_.values[i] * a_.values[i];
+      else
+        r_.values[i] += b_.values[i] * a_.values[i-1];
+    }
+
+    return simde_float32x2_from_private(r_);
+   #endif
+ }
+ #if defined(SIMDE_ARM_NEON_A32V8_ENABLE_NATIVE_ALIASES)
+   #undef vcmla_f32
+   #define vcmla_f32(r, a, b) simde_vcmla_f32(r, a, b)
+ #endif
+
+ SIMDE_FUNCTION_ATTRIBUTES
+ simde_float32x4_t
+ simde_vcmlaq_f32(simde_float32x4_t r, simde_float32x4_t a, simde_float32x4_t b) {
+   #if defined(SIMDE_ARM_NEON_A32V8_NATIVE)
+    return vcmlaq_f32(r, a, b);
+   #else
+    simde_float32x4_private
+      r_ = simde_float32x4_to_private(r),
+      a_ = simde_float32x4_to_private(a),
+      b_ = simde_float32x4_to_private(b);
+
```

Comment:
This can be better implemented with a shuffle.  Clang's documentation explains the concept: https://clang.llvm.org/docs/LanguageExtensions.html#langext-builtin-shufflevector fairly well.

In SIMDe, we have a `SIMDE_SHUFFLE_VECTOR_` macro which abstracts away the differences between clang's `__builtin_shufflevector` and GCC's `__builtin_shuffle`.  It's not defined for compilers which don't implement the required functionality, so you can test `defined(SIMDE_SHUFFLE_VECTOR_)` to see if you should use it.  The first argument is the width in bits of each element, the second is the size of the entire vector in bytes, then the two input vectors (which are often the same if you only want to shuffle lanes within a single vector), followed by the lane numbers to copy.

In this case, the values for `a` that we want to use are `{ 0, 0, 2, 2 }`, so we can do something like this:

```c
  #if defined(SIMDE_SHUFFLE_VECTOR_)
    a_.values = SIMDE_SHUFFLE_VECTOR_(32, 16, a_.values, a_.values, 0, 0, 2, 2);
    r_.values += b_.values * a_.values;
  #else
    SIMDE_VECTORIZE
    ...
```

Both GCC and clang are very good at generating optimal code for shuffles, so that basically takes care of all the complexity of this function.  After that, all we have is a simple multiply and add.

Reply:
This is pretty cool! Thanks, I'll add an implementation using shuffles.